src/rust/engine/process_execution/src/bounded_tests.rs

```diff
@@ -327,3 +332,43 @@ async fn dropped_future_is_removed_from_queue() {
   }
   assert_eq!(1, sema.available_permits());
 }
+
+/// Given Tasks as triples of desired, actual, and expected concurrency (all of which are
+/// assumed to be preemptible), assert that the expected concurrency is applied.
+fn test_balance(
+  concurrency_limit: usize,
+  expected_preempted: usize,
+  task_defs: Vec<(usize, usize, usize)>,
+) {
+  let ten_minutes_from_now = Instant::now() + Duration::from_secs(10 * 60);
+  let tasks = task_defs
+    .iter()
+    .enumerate()
+    .map(|(id, (desired, actual, _))| Task::new(id, *desired, *actual, ten_minutes_from_now))
+    .collect::<Vec<_>>();
+
+  assert_eq!(
+    expected_preempted,
+    balance(concurrency_limit, Instant::now(), tasks.iter().collect())
+  );
+  for (task, (_, _, expected)) in tasks.iter().zip(task_defs.into_iter()) {
+    assert_eq!(expected, task.concurrency());
+  }
+}
+
+#[tokio::test]
+async fn balance_noop() {
+  test_balance(2, 0, vec![(1, 1, 1), (1, 1, 1)]);
+}
+
+#[tokio::test]
+async fn balance_overcommitted() {
+  // Preempt the first Task and give it one slot, without adjusting the second task.
+  test_balance(2, 1, vec![(2, 2, 1), (1, 1, 1)]);
```

Comment:
Would this be right?

```rust
test_balance(1, 1, vec![(1, 1, 1), (1, 1, 0)]);
```

If so, maybe worth adding.

Reply:
> Would this be right?

It wouldn't be, because we will never give a process less than 1 concurrency slot. And because we only calculate concurrency once we have a slot on the semaphore, having more processes than there is concurrency available (i.e. `#processes > #permits`) shouldn't be possible.